src/coingecko/Coingecko.ts

```diff
@@ -275,6 +270,58 @@ export class Coingecko {
     return this.prices[platform_id][currency];
   }
 
+  protected getCachedAddressPrice(
+    contractAddress: string,
+    currency: string,
+    platform_id: string
+  ): CoinGeckoPrice | undefined {
+    const priceCache = this.getPriceCache(currency, platform_id);
+    const now: number = msToS(Date.now());
+    const tokenPrice: CoinGeckoPrice | undefined = priceCache[contractAddress];
+    if (tokenPrice === undefined || tokenPrice.timestamp + this.maxPriceAge <= now) {
+      if (this.maxPriceAge > 0) {
+        this.logger.debug({
+          at: "Coingecko#getCurrentPriceByContract",
+          message: `Cache miss on ${platform_id}/${currency} for ${contractAddress}`,
+          maxPriceAge: this.maxPriceAge,
+          tokenPrice: tokenPrice,
+        });
+      }
+      return undefined;
+    } else {
+      this.logger.debug({
+        at: "Coingecko#getCurrentPriceByContract",
+        message: `Cache hit on token ${contractAddress} (age ${now - tokenPrice.timestamp} S).`,
+        price: tokenPrice,
+      });
+      return tokenPrice;
+    }
+  }
+
+  protected updatePriceCache(cgPrice: CGTokenPrice, contractAddress: string, currency: string, platform_id: string) {
+    const priceCache = this.getPriceCache(currency, platform_id);
+    if (priceCache[contractAddress] === undefined) {
+      priceCache[contractAddress] = { address: contractAddress, price: 0, timestamp: 0 };
+    }
+    if (cgPrice.last_updated_at > priceCache[contractAddress].timestamp) {
+      priceCache[contractAddress] = {
+        address: contractAddress,
+        price: cgPrice[currency],
+        timestamp: cgPrice.last_updated_at,
+      };
+      this.logger.debug({
+        at: "Coingecko#updatePriceCache",
+        message: `Updated ${platform_id}/${currency}/${contractAddress} token price cache.`,
+      });
+    } else if (cgPrice.last_updated_at === priceCache[contractAddress].timestamp) {
```

Comment:
OOC would we want to relax this constraint? If we make it `else` then we'd get the case where we pass an earlier value
```suggestion
    } else {
```

Reply:
Yes, we can do that. I would expect that case to be uncommon but can def use just else here.